src/type.ml

```diff
@@ -674,10 +821,105 @@ and tag_typ_constr
     let tag_typs = List.combine typs should_tag_list in
     Monad.List.map (fun (typ, should_tag) ->
         if should_tag
-        then tag_typ_constr_aux typ
+        then tag_typ_constr_aux existencial_typs typ
         else return typ
       ) tag_typs
 
+and record_args
+  (labeled_typs : Types.label_declaration list)
+  : (t list * VarEnv.t) Monad.t =
+  let* (record_params, _, new_typ_vars) = labeled_typs |>
+    List.map (fun { Types.ld_type; _ } -> ld_type) |>
+    (fun ls -> of_typs_exprs true ls Name.Map.empty)
+  in
+  let typ_args =
+    List.fold_left
+      (fun typ_args new_typ_args ->
+         (typ_args @
+          Name.Set.elements (
+            Name.Set.diff new_typ_args (Name.Set.of_list typ_args)
+          ))
+      )
+      []
+      (List.map typ_args_of_typ record_params) in
+  let new_typ_vars = VarEnv.reorg typ_args new_typ_vars in
+  return (record_params, new_typ_vars)
+
+
+and typed_existential_typs_of_typ
```

Comment:
We a comment for what does this function do (I should comment more my code too).